Allow markdown formatting in publisher statement

awead opened the pull request:
We need to join the two fields together with the correct combination of line feed and carriage return so that it will all parse as one string.

awead commented on app/services/analytics/view_report.rb:
```diff
@@ -0,0 +1,32 @@
+    private
+
+      def google_analytics_view_results
+        return @google_analytics_view_results if @google_analytics_view_results.present?
```
I took me a little while to figure out you're memoizing this. Is this better?
``` ruby
def google_analytics_view_results
  @google_analytics_view_results ||= init_view_results
end

def init_view_results
  results = Sufia::Pageview.results(Sufia::Analytics.profile, start_date: start_date, end_date: end_date, sort: :pagePath)
  results.dimensions << :pagePath
  results
end
```

cam156 replied:
@awead I do not know.  I had similar code to that and I removed it becuase the top method seemed silly.

awead commented on app/services/analytics/download_report.rb:
```diff
@@ -0,0 +1,69 @@
+
+    def work_downloads
+      initialize_page_downloads
+      @work_downloads
```
Can this be: `@work_downloads ||= initialize_page_downloads`

cam156 replied:
No downloads is both File and Work downloads.  We could run two queries and just look for files the first time and Works the second time.

awead replied:
That makes more sense to me.

cam156 replied:
@awead so you really want to send the query to google twice.  That seems unfortunate.

awead commented on app/services/analytics/download_report.rb:
```diff
@@ -0,0 +1,69 @@
+
+    def file_set_downloads
+      initialize_page_downloads
+      @file_set_downloads
```
same here: `@file_set_downloads ||= initialize_page_downloads`

cam156 replied:
@awead no same as above how could the same method return two different things?

awead commented on app/services/analytics/download_report.rb:
```diff
@@ -0,0 +1,69 @@
+      end
+
+      def initialize_page_downloads
+        return if @work_downloads.present?
```
you shouldn't need this is the above change can be made

awead commented on app/services/analytics/download_report.rb:
```diff
@@ -0,0 +1,69 @@
+        @work_downloads = []
+        @file_set_downloads = []
+        google_analytics_download_results.each do |page_download|
+          @work_downloads, @file_set_downloads = classify_download(page_download, @work_downloads, @file_set_downloads)
```
I feel like if you could put `classify_download` within thie each loop, we would need to pass around the `work_downloads` and `file_set_downloads` variables. It's confusing to read.

cam156 replied:
@awead I do not understand this comment. The arrays are in the each loop.  They are being passed in and returned.  We could just use the instance variables, but you are arguing against them above.

awead commented on app/services/analytics/download_report.rb:
```diff
@@ -0,0 +1,69 @@
+      end
+
+      def lookup_object(id)
+        @last_object ||= GenericWork.new
```
It isn't immediately obvious to me what `last_object` is until you trace it back up through `classify_download` and back into the each loop at line 40. Is there a way we can clarify that a bit more?

cam156 replied:
@awead the last object was the last object pulled out of fedora.  What would you have the name be?  The last_object_loaded_from_fedora ???

awead commented on app/services/user_stats_importer.rb:
```diff
@@ -0,0 +1,83 @@
+  end
+
+  def tally_user_results
+    sorted_users.each do |user|
```
where is sorted_users created?

cam156 replied:
@awead that is a part of the parent class.

awead replied:
@cam156 I think that needs to be documented then.

cam156 replied:
@awead You want to document every use of object oriented inheritance in the code?

cam156 replied:
If I am hearing correctly any time an object uses a method from the parent object it should be documented, so every time we use find from active fedora we should add a comment that says this is a method inherited from ActiveFedora and is not defined as part of our model, but part of the AftiveFedora::Base

cam156 commented on spec/services/download_report_spec.rb:
```diff
@@ -0,0 +1,90 @@
+      allow(generic_file_with_fileset).to receive(:file_sets).and_return([FileSet.new(id: '3xs55m950_file', depositor: user2.login)])
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m950').twice.and_return(generic_file_with_fileset)
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m9505').once.and_return(GenericWork.new(id: '3xs55m9505', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10').once.and_return(GenericWork.new(id: '1vh53wt10', depositor: user2.login))
```
https://github.com/psu-stewardship/scholarsphere/pull/1028/files#diff-d4425a6b2765b7eb294f239cf6707a6dR42

cam156 commented on spec/services/download_report_spec.rb:
```diff
@@ -0,0 +1,90 @@
+      allow(ActiveFedora::Base).to receive(:find).with('fj67314220').once.and_return(FileSet.new(id: 'fj67314220', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('nzs25x853x').once.and_return(FileSet.new(id: 'nzs25x853x', depositor: user2.login))
+      allow(ActiveFedora::Base).to receive(:find).with('4f4752g26g').once.and_return(FileSet.new(id: '4f4752g26g', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('other_missing').and_raise(ActiveFedora::ObjectNotFoundError)
```
@awead the code does look for the exception here: https://github.com/psu-stewardship/scholarsphere/pull/1028/files#diff-366be95a4065da0fa459fd2ae2abaa74R52

cam156 commented on spec/services/download_report_spec.rb:
```diff
@@ -0,0 +1,90 @@
+      allow(report).to receive(:google_analytics_download_results).and_return(download_results)
+      allow(generic_file_with_fileset).to receive(:file_sets).and_return([FileSet.new(id: '3xs55m950_file', depositor: user2.login)])
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m950').twice.and_return(generic_file_with_fileset)
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m9505').once.and_return(GenericWork.new(id: '3xs55m9505', depositor: user1.login))
```
@awead I will have to look. They may be left over from when all the tests were together.

cam156 commented:
@awead Hmm.  I seem to have missed on my Naming.  ViewReport grabs a report of all the views from Google Analytics and parses it.  Download Report does similar for downloads.  Let's start at the top.  What would you name those so you can understand what they do?

cam156 commented:
@awead based on your comments I assumed the code was completely off.  I spent the morning doing some refactoring and renaming.  Maybe it makes a bit more sense now?

awead commented:
@cam156 Yes. I think it exposes things that are in need of refactoring later, but some of those are really coming from the super class. I tried my hand at taming this beast and couldn't do any better, so I'd say let just merge it and move on.

awead committed:
Allow markdown formatting in publisher statement

We need to join the two fields together with the correct combination of
line feed and carriage return so that it will all parse as one string.
